Utils/src/FastSimWeightPR31285To36122.cc

```diff
@@ -0,0 +1,198 @@
+   hRatio_GenJetHadronPtGenJetHadronFlavorEqEq5->SetBinContent(17,5.119048);
+   hRatio_GenJetHadronPtGenJetHadronFlavorEqEq5->SetBinContent(18,5.574074);
+   hRatio_GenJetHadronPtGenJetHadronFlavorEqEq5->SetBinContent(19,6.516129);
+   hRatio_GenJetHadronPtGenJetHadronFlavorEqEq5->SetBinContent(20,7.685393);  
+}
+// ------------ method called for each event  ------------
+void FastSimWeightPR31285To36122::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+    using namespace edm;
+    double EventWeight = 1.0;    
+    edm::Handle< View<reco::GenParticle> > genPartCands;
+    iEvent.getByToken(genCollectionTok, genPartCands);
+    edm::Handle< View<reco::GenJet> > genJets;
+    iEvent.getByToken(genJetTok, genJets);
+    edm::Handle< View<pat::Jet> > recJets;
+    iEvent.getByToken(recJetTok, recJets);   
+    bool jetHasOffendingGp_;
+    float dr;    
+    float mindr;
+    double leadHadronPt;
+    float originxy;   
+    float decayxy;       
+    int hadronFlavor;     
+    int pdgid;
+    const float genJetPtThreshold = 30;
+    const float genParticlePtThreshold = 15;
+    const float drThreshold = 0.4;
+    const float justInsidePipe = 2.16;
+    const float justOutsidePipe = 2.17;
+    const float farOutsideCaloButNotTooFarAway = 2000;
+    for(const auto& genJet : *genJets) 
+    {
+           if (!(genJet.pt()>genJetPtThreshold)) continue;
+           jetHasOffendingGp_ = false;//must use
+           mindr = 99;
+           hadronFlavor = -11;
+           leadHadronPt = 1;
+           originxy = -1;
+           for(const auto& recJet : *recJets) 
+            {
+                dr = deltaR(recJet.p4(),genJet.p4());
+                if (!(dr<drThreshold && dr<mindr)) continue;
+                mindr = dr;
+                hadronFlavor = std::max(1,recJet.hadronFlavour());
+                for(const auto& iPart : *genPartCands) {
+                    if (!(iPart.pt()>genParticlePtThreshold)) continue;
+                    pdgid = abs(iPart.pdgId());
+                    if (!(deltaR(iPart,genJet)<drThreshold)) continue;
+                    if(!(pdgid>100 && pdgid<600)) continue; //only consider hadrons
+                    if (iPart.pt()>leadHadronPt) {
+                        if ( (hadronFlavor<4 && pdgid>100 && pdgid<400) ||//only consider hadrons
+                             (hadronFlavor==4 && pdgid>400 && pdgid<500) ||//only consider hadrons
+                             (hadronFlavor==5 && pdgid>400 && pdgid<600) )//only consider hadrons
+                        {
+                            leadHadronPt = iPart.pt();
+                        }
+                    }                       
+                    if(!(iPart.numberOfDaughters()>1)) continue;
+                    originxy = sqrt(pow(iPart.vx(),2) + pow(iPart.vy(),2));
+                    if(!(originxy<justInsidePipe)) continue;
+                    decayxy = sqrt(pow(iPart.daughter(0)->vx(),2) + pow(iPart.daughter(0)->vy(),2));
```

Comment:
it would be more efficient to avoid the `sqrt()` by making the constants used below into the squared versions, e.g. `const float justOutsidePipe = 2.17*2.17;`

Reply:
done